src/runtime_src/core/tools/xbmgmt2/SubCmdFlash.cpp

```diff
@@ -163,17 +233,159 @@ bool canProceed()
   return proceed;
 }
 
+/* 
+ * Flash shell and sc firmware
+ * Helper method for auto_flash
+ */
+static int 
+updateShellAndSC(uint16_t boardIdx, DSAInfo& candidate, bool& reboot)
+{
+  reboot = false;
+
+  Flasher flasher(boardIdx);
+  if(!flasher.isValid()) {
+    std::cout << "card not available" << std::endl;
+    return -EINVAL;
+  }
+
+  bool same_dsa = false;
+  bool same_bmc = false;
+  DSAInfo current = flasher.getOnBoardDSA();
+  if (!current.name.empty()) {
+    same_dsa = (candidate.name == current.name &&
+      candidate.matchId(current));
+    same_bmc = (current.bmcVer.empty() ||
+      candidate.bmcVer == current.bmcVer);
+  }
+  if (same_dsa && same_bmc)
+    std::cout << "update not needed" << std::endl;
+
+  if (!same_bmc) {
+    std::cout << "Updating SC firmware on card[" << flasher.sGetDBDF() <<
+      "]" << std::endl;
+    auto ret = 0;update_SC(boardIdx, candidate.file.c_str());
+    if (ret != 0) {
+      std::cout << "WARNING: Failed to update SC firmware on card ["
+        << flasher.sGetDBDF() << "]" << std::endl;
+    }
+  }
+
+  if (!same_dsa) {
+    std::cout << "Updating shell on card[" << flasher.sGetDBDF() <<
+      "]" << std::endl;
+    auto ret = 0;update_shell(boardIdx, "", candidate.file.c_str(),
+      candidate.file.c_str());
+    if (ret != 0) {
+      std::cout << "ERROR: Failed to update shell on card["
+        << flasher.sGetDBDF() << "]" << std::endl;
+    } else {
+      reboot = true;
+    }
+  }
+
+  if (!same_dsa && !reboot)
+    return -EINVAL;
+
+  return 0;
+}
+
+/* 
+ * Update shell and sc firmware on the card automatically
+ */
 static void 
 auto_flash(uint16_t index, std::string& name,
     std::string& id, bool force) 
 {
-  //to-do
-  index = index;
-  name = name;
-  id = id;
-  force = force;
+  std::vector<uint16_t> boardsToCheck;
+  std::vector<std::pair<uint16_t, DSAInfo>> boardsToUpdate;
+
+  // Sanity check input dsa and timestamp.
+  if (!name.empty()) {
+        bool foundDSA = false;
+    bool multiDSA = false;
+    auto installedDSAs = firmwareImage::getIntalledDSAs();
+    for (DSAInfo& dsa : installedDSAs) {
+      if (name == dsa.name &&
+        (id.empty() || dsa.matchId(id))) {
+        if (!foundDSA)

```

Comment:
Wouldn't it be better to do the following:
```
multiDSA = multiDSA || foundDSA;
foundDSA = true;

```

Reply:
Hmm. I don't quite understand this.

Reply:
```
if (!x) 
  x=true;
else 
 y=true
```
same as 
```
y=y || x;
x=true
```

Reply:
Thanks. That helped :)